ayanda/src/main/java/sintulabs/p2p/Bluetooth.java

```diff
@@ -0,0 +1,240 @@
+package sintulabs.p2p;
+
+import android.app.Activity;
+import android.bluetooth.BluetoothAdapter;
+import android.bluetooth.BluetoothDevice;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.LocalBroadcastManager;
+import android.util.Log;
+import android.widget.Toast;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static android.bluetooth.BluetoothAdapter.ACTION_DISCOVERY_FINISHED;
+import static android.bluetooth.BluetoothAdapter.ACTION_DISCOVERY_STARTED;
+import static android.bluetooth.BluetoothAdapter.ACTION_SCAN_MODE_CHANGED;
+import static android.bluetooth.BluetoothAdapter.ACTION_STATE_CHANGED;
+import static android.bluetooth.BluetoothAdapter.SCAN_MODE_CONNECTABLE;
+import static android.bluetooth.BluetoothAdapter.SCAN_MODE_CONNECTABLE_DISCOVERABLE;
+import static android.bluetooth.BluetoothAdapter.STATE_ON;
+
+
+/**
+ * Created by sabzo on 1/14/18.
+ */
+
+public class Bluetooth extends P2P {
+    private Context context;
+    BluetoothAdapter mBluetoothAdapter;
+    private BroadcastReceiver receiver;
+    private IntentFilter intentFilter;
+    public static String BT_DEVICE_FOUND = "4000";
+    public static Integer REQUEST_ENABLE_BT = 1;
+    public static Integer BT_PERMISSION_REQUEST_LOCATION = 4444;
+    public static Integer BT_ENABLED = 3000;
+    private Boolean discoveryInitiated = false;
+    private Set<String> deviceNamesDiscovered;
+    private List<Device> deviceList;
+
+    private IBluetooth iBluetooth;
+
+
+    public Bluetooth(Context context, IBluetooth iBluetooth) {
+        this.context = context;
+        this.iBluetooth = iBluetooth;
+        mBluetoothAdapter= BluetoothAdapter.getDefaultAdapter();
+        deviceNamesDiscovered = new HashSet<>();
+        deviceList = new ArrayList<>();
+        createIntentFilter();
+        createReceiver();
+        // ensure to register and unregister receivers
+    }
+
+    @Override
+    public Boolean isSupported() {
+       return  (mBluetoothAdapter == null)? false : true;
+    }
+
+    @Override
+    public Boolean isEnabled() {
+        return mBluetoothAdapter.isEnabled();
+    }
+
+    private void enable() {
+        Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
+        ((Activity)context).startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
```

Comment:
Better to return an `Intent` instead of starting it. This way the consumer can react in `onActivityResult`.